platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/mappingToJvm/RuntimeFunctionResolver.kt

```diff
@@ -42,48 +46,103 @@ interface RuntimeFunctionResolver {
-            // It is not very nice that we rely on this internal signature, but we don't have much of a choice...
-            // We have also tried using the "descriptor" field and the "overriddenDescriptors" and/or "overriddenFunctions" from that,
-            // but it's even uglier with a lot of weird lazy types involved and the code for doing so becomes very brittle.
+object DefaultRuntimeFunctionCandidatesProvider : RuntimeFunctionCandidatesProvider {
+    override fun candidatesForMember(
+        ownerKClass: KClass<*>,
+        schemaFunction: SchemaMemberFunction
+    ): List<KFunction<*>> = ownerKClass.memberFunctions.filter { isMatchByNameAndAnnotations(schemaFunction, ownerKClass, it) }
+
+    override fun candidatesForTopLevelFunction(schemaFunction: DataTopLevelFunction, scopeClassLoader: ClassLoader): List<KFunction<*>> {
+        val ownerClass = try {
+            scopeClassLoader.loadClass(schemaFunction.ownerJvmTypeName)
+        } catch (_: ClassNotFoundException) {
+            return emptyList()
         }
 
-        fun matchesAnnotationsRecursively(function: KFunction<*>, receiverClass: KClass<*>, predicate: (Annotation) -> Boolean): Boolean =
-            if (function.annotations.any(predicate)) {
-                true
-            } else {
-                val functionSignature = signature(function)
-                receiverClass.allSuperclasses.any { parentClass ->
-                    val parentFunction = parentClass.memberFunctions.firstOrNull { signature(it) == functionSignature }
-                    parentFunction?.annotations?.any(predicate) ?: false
-                }
+        return ownerClass.methods.mapNotNull {
+            if (it.name == schemaFunction.simpleName)
+                it.kotlinFunction?.takeIf { kotlinFunction -> isMatchByNameAndAnnotations(schemaFunction, ownerClass.kotlin, kotlinFunction) }
+            else null
+        }
+    }
+
+    private fun isMatchByNameAndAnnotations(
+        schemaFunction: SchemaFunction,
+        ownerKClass: KClass<*>,
+        kFunction: KFunction<*>,
+    ): Boolean = kFunction.name == schemaFunction.simpleName &&
+        (schemaFunction is DataTopLevelFunction ||
+            matchesAnnotationsRecursively(kFunction, ownerKClass, hasDeclarativeAnnotation))
+
+    private fun matchesAnnotationsRecursively(function: KFunction<*>, receiverClass: KClass<*>, predicate: (Annotation) -> Boolean): Boolean =
+        if (function.annotations.any(predicate)) {
+            true
+        } else {
+            val functionSignature = signature(function)
+            receiverClass.allSuperclasses.any { parentClass ->
+                val parentFunction = parentClass.memberFunctions.firstOrNull { signature(it) == functionSignature }
+                parentFunction?.annotations?.any(predicate) ?: false
             }
+        }
+
+    private fun signature(function: KFunction<*>): String {
```

Comment:
Maybe it would be worth putting this into `platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/workarounds.kt` so that we might remember it and replace it in the future.

Reply:
Done in [`365f734` (#32409)](https://github.com/gradle/gradle/pull/32409/commits/365f7343c63122ca1d5ae5fe88d6c7d1c14c4baa)